src/measure_theory/interval_integral.lean

```diff
@@ -1582,4 +1583,30 @@ begin
   { exact (hcv.mul hcu').interval_integrable },
 end
 
+/-!
+### Integration by substitution / Change of variables
+-/
+
+theorem integral_comp_mul_deriv' {f f' g : ℝ → ℝ}
+  (hf : ∀ x ∈ interval a b, has_deriv_at f (f' x) x)
+  (hf' : continuous_on f' (interval a b))
+  (hg : ∀ x ∈ f '' (interval a b), continuous_at g x)
+  (hgm : ∀ x ∈ f '' (interval a b), measurable_at_filter g (𝓝 x)) :
```

Comment:
Doesn't `hg` imply `hgm`? If `g` is continuous at `x`, it is continuous on a small neighborhood of `x` so it is ae-measurable w.r.t. the measure restricted to that neighborhood, so it is measurable at the filter `𝓝 x`?

Reply:
I know a way to prove this (see below) but it requires that `f '' (interval a b)` is open, which I don't think we have? Is there something I'm missing here?
```
import measure_theory.interval_integral

open_locale topological_space

example {a b : ℝ} {f g : ℝ → ℝ} {s : set ℝ} (hs : is_open s) (hg : ∀ x ∈ s, continuous_at g x) :
  ∀ x ∈ s, measurable_at_filter g (𝓝 x) :=
λ x hx, ⟨s, mem_nhds_sets hs hx, (continuous_at.continuous_on hg).ae_measurable hs.measurable_set⟩
```

Reply:
> If `g` is continuous at `x`, it is continuous on a small neighborhood of `x` so...

This claim by Floris is wrong. There are functions continuous at only one point. Here `f` could be the constant function mapping every real number to 0, and g could send x to itself if x is rational and zero otherwise. In this example f is smooth everywhere, g is continuous at every point of the image of x but not on a neighborhood of the image of f. Of course g is measurable since it is explicit, but at least this means Floris' easy argument doesn't work.

Reply:
@PatrickMassot Are you saying that I do need to assume `hgm`, or just that it can't be derived via the method that Floris suggested?

Also, out of curiosity, is there any significance to the result I showed above (i.e. for an arbitrary open set `s`) or is it trivial?

Reply:
Thanks Patrick, this proof indeed doesn't work. I think your counterexample can be adapted to a counterexample of `continuous_at g x ->  measurable_at_filter g (𝓝 x)`.
Benjamin, I think your lemma for open set `s` is useful to have in mathlib. It is not applicable here, since ` f '' (interval a b)` is never open in this theorem.
I'm quite sure that also means that `hg` does not imply `hgm`.

It should still be possible to remove `hgm` from this particular theorem, but it requires generalizing some previous lemmas. `hf` (and in the proof `h`) only have to be true for `x` in the interior of `interval a b`. If you do that, `hgm` is only needed for `x` in the interior of `interval a b`, which can be proven the lemma benjamin mentions. But that means that a lemma like `integral_eq_sub_of_has_deriv_at` has to be generalized as well.
I suggest we accept this PR as-is and try to weaken the hypotheses in a later PR.

Reply:
Oh wait, nevermind: can you use `integral_eq_sub_of_has_deriv_at'` instead of `integral_eq_sub_of_has_deriv_at`? That should be able to weaken `hf` and remove `hgm`.

Reply:
Sure!

Reply:
In that case you will have to add the assumption that `f` is continuous on `interval a b` (which can be removed in the corollary where you assume that `f` is differentiable on all of `interval a b`).

Also, is there a reason why you write `∀ x ∈ f '' (interval a b), continuous_at g x` instead of using `continuous_on` like in most of this file?

Reply:
Don't I assume that `f` is differentiable on all of `interval a b` here as well?

And IIRC the proof has one instance where it needed it to be`continuous_at`, but perhaps I can switch that up.

Reply:
> Don't I assume that f is differentiable on all of interval a b here as well?

I'm asking you to change that.

Reply:
Oh yes of course

Reply:
I've run into an issue: `integral_eq_sub_of_has_deriv_at'` requires `continuous_on (λ (u : ℝ), ∫ (t : ℝ) in f a..f u, g t) (interval a b)`. Since we don't yet have the fact that integrals are continuous I don't think this is attainable without making further assumptions. Is that correct?

Reply:
Hmm... You're saying that we don't have the fact that an integral of a measurable is continuous yet? That is unfortunate. We do have that the integral is continuous on the interior of the interval (because it is differentiable there). 

We should really prove that the integral is continuous soon, since we're now doing things out of order (we have to add much stronger hypotheses to lemmas than needed). Would you be interested in trying to prove that the integral of a measurable function (on a set) is continuous (on that set)? It shouldn't be too hard. E.g. here is a proof sketch:
https://math.stackexchange.com/questions/1477452/the-lebesgue-integral-of-an-integrable-function-is-continuous

I would prefer if "integrals are continuous" is proven before this PR is merged, but I could also live with merging this PR and weakening the assumptions later (FTC-2 also has unnecessarily strong assumptions IIRC). In that case we should add a  note.

Reply:
We have `interval_integral.continuous_on_integral_of_continuous` but that would require me to assume that `g` is `continuous` and even then I'm not sure that they're reconcilable.

I am definitely willing to undertake trying to prove "integrals are continuous." Having said that, I don't know how long it will take me, and I have a few projects (unaffected by this issue) that depend on the content of this PR, so I prefer to have this merged and then revisit here and FTC-2 once "integrals are continuous" is proved. Does that sound alright?

Reply:
We don't want to assume that `g` is continuous everywhere.
`interval_integral.continuous_on_integral_of_continuous` should be proven without the condition `continuous f`.

OK, I'm fine with merging this. I hope that your projects that depend on this are not also stated with overly strict conditions (we are claiming for 4 months already that we have proven FTC-2 in `undergrad.yaml` and `100.yaml` while we only have a special case of it, and it would be nice to see that fixed).

Reply:
> We don't want to assume that `g` is continuous everywhere.
> `interval_integral.continuous_on_integral_of_continuous` should be proven without the condition `continuous f`.

Okay, that's exactly what I had figured.

> OK, I'm fine with merging this. I hope that your projects that depend on this are not also stated with overly strict conditions (we are claiming for 4 months already that we have proven FTC-2 in `undergrad.yaml` and `100.yaml` while we only have a special case of it, and it would be nice to see that fixed).

No, the projects I am referring to all involve continuous functions so they use the other lemma I add here anyway and are unaffected by this discussion.

Reply:
So should I add a TODO and then we can merge this?

Reply:
Sorry, I forgot to responds. Yes, if you add a TODO I'm happy to merge

Reply:
No worries. This PR slipped my mind as well. I've added the TODO.